libmcount/trigger.c

```diff
@@ -0,0 +1,114 @@
+#include <fcntl.h>
+#include <stdbool.h>
+#include <inttypes.h>
+
+#include "libmcount/mcount.h"
+#include "utils/env-file.h"
+#include "utils/utils.h"
+
+
+#define ERROR_LIBRARY_NOT_EXIST \
+"%s does not exist libmcount shared object.\n"				\
+"\tmake sure that the file exists.\n\n"					\
+"\t[notification]\n"							\
+"\twhen you use dynamic tracing to the process, you must specify the\n"	\
+"\tabsolute path of 'libmcount-dynamic.so' not the relative path.\n\n" 	\
+"\tfor example:\n" 							\
+"\t[BADD]./uftrace dynamic -L. --pid=xxxx\n"				\
+"\t[GOOD]./uftrace dynaimc -L/home/m/uftrace  --pid=xxxx\n"
+
+#define ERROR_DATADIR_NOT_EXIST \
+"Cannot access to the %s for saving uftrace datas.\n"  			\
+"\tmake sure that the path exists.\n\n" 				\
+"\t[notification]\n"							\
+"\twhen you use dynamic tracing to the process, you must specify the\n"	\
+"\tabsolute path of data-dir.\n\n" 					\
+"\tfor example:\n" 							\
+"\t[BADD]./uftrace dynamic -L/home/m/uftrace --pid=xxxx\n" 		\
+"\t[GOOD]./uftrace dynaimc -L/home/m/uftrace"				\
+"--data=/home/m/uftrace/data --pid=xxxx\n"
+
+#define ERROR_ENVFILE_NOT_EXIST \
+"\tCannot access to the %s which contain environment variable.\n"
+
+
+__attribute__((weak)) void build_debug_domain(char *dbg_domain_str)
+{
+	int i, len;
+
+	if (dbg_domain_str == NULL)
+		return;
+
+	len = strlen(dbg_domain_str);
+	for (i = 0; i < len; i += 2) {
+		const char *pos;
+		char domain = dbg_domain_str[i];
+		int level = dbg_domain_str[i+1] - '0';
+		int d;
+
+		pos = strchr(DBG_DOMAIN_STR, domain);
+		if (pos == NULL)
+			continue;
+
+		d = pos - DBG_DOMAIN_STR;
+		dbg_domain[d] = level;
+	}
+}
+
+__attribute__((constructor)) void so_main(void)
+{
+	char *debug_str, *preload_library_path, *data_dir_path;
+	int fd_envfile;
+	struct stat file;
+	void *handle;
+
+	outfp = stdout;
+	logfp = stderr;
+
+	/*
+	 * the Process output is delayed for unknown reasons when using
+	 * dynamic tracing. cannot found the reason of delay but below code
+	 * can mitigation the symptom.
+	 */
+	setvbuf(stdout, NULL, _IONBF, 1024);
+	setvbuf(stderr, NULL, _IONBF, 1024);
+
+	fd_envfile = open_env_file();
+	if (fd_envfile < 0)
+		printf(ERROR_ENVFILE_NOT_EXIST, ENV_FILE);
+
+	set_env_from_file(fd_envfile);
+	preload_library_path = getenv("LD_PRELOAD");
+	data_dir_path = getenv("UFTRACE_DIR");
+	debug_str = getenv("UFTRACE_DEBUG");
+	if (debug_str) {
+		debug = strtol(debug_str, NULL, 0);
+		build_debug_domain(getenv("UFTRACE_DEBUG_DOMAIN"));
+	}
+
+	pr_dbg("LIBRARY PATH : %s\n", preload_library_path);
+	pr_dbg("DATADIR PATH : %s\n", data_dir_path);
+
+	if (stat(data_dir_path, &file) == 0)
+		pr_dbg("DATA-DIR EXIST : %s\n", data_dir_path);
+	else
+		pr_err_ns(ERROR_DATADIR_NOT_EXIST, data_dir_path);
+
+	if (stat(preload_library_path, &file) == 0)
+		pr_dbg("LIBRARY EXIST : %s\n", preload_library_path);
+	else
+		pr_err_ns(ERROR_LIBRARY_NOT_EXIST, preload_library_path);
+
+	handle = dlopen(preload_library_path, RTLD_LAZY);
```

Comment:
Hmm.. so do you want to load libmcount.so again?  Why not just doing it in this library?

Reply:
because libmcount could not loaded by using __libc_dlopen_mode but it can be possible by using dlopen. I don't know what cause this. see below backtrace. 
```
   0x000000000040062a <+19>:    callq  0x400520 <__libc_dlopen_mode@plt>
   0x000000000040062f <+24>:    mov    $0x40071c,%edi
   0x0000000000400634 <+29>:    callq  0x4004f0 <puts@plt>
   0x0000000000400639 <+34>:    mov    0x200458(%rip),%rax        # 0x600a98 <so_name>
   0x0000000000400640 <+41>:    mov    $0x2,%esi
   0x0000000000400645 <+46>:    mov    %rax,%rdi
   0x0000000000400648 <+49>:    callq  0x400510 <dlopen@plt>
   0x000000000040064d <+54>:    mov    $0x0,%eax
   0x0000000000400652 <+59>:    pop    %rbp
   0x0000000000400653 <+60>:    retq
End of assembler dump.
(gdb) b *main+49
Breakpoint 1 at 0x400648
(gdb) r
Starting program: /home/m/git/a.out

Program received signal SIGSEGV, Segmentation fault.
_dl_map_object_from_fd (name=name@entry=0x4006e8 "/home/m/git/_uftrace/libmcount/libmcount-dynamic.so", origname=origname@entry=0x0,
    fd=<optimized out>, fbp=fbp@entry=0x7fffffffdba0, realname=<optimized out>, loader=loader@entry=0x0, l_type=2, mode=268435458,
    stack_endp=0x7fffffffdb98, nsid=0) at dl-load.c:1337
1337    dl-load.c: No such file or directory.
(gdb) bt
#0  _dl_map_object_from_fd (name=name@entry=0x4006e8 "/home/m/git/_uftrace/libmcount/libmcount-dynamic.so", origname=origname@entry=0x0,
    fd=<optimized out>, fbp=fbp@entry=0x7fffffffdba0, realname=<optimized out>, loader=loader@entry=0x0, l_type=2, mode=268435458,
    stack_endp=0x7fffffffdb98, nsid=0) at dl-load.c:1337
#1  0x00007ffff7ddfc27 in _dl_map_object (loader=0x0, loader@entry=0x7ffff7ffe168,
    name=name@entry=0x4006e8 "/home/m/git/_uftrace/libmcount/libmcount-dynamic.so", type=type@entry=2, trace_mode=trace_mode@entry=0,
    mode=mode@entry=268435458, nsid=<optimized out>) at dl-load.c:2498
#2  0x00007ffff7dec577 in dl_open_worker (a=a@entry=0x7fffffffe160) at dl-open.c:237
#3  0x00007ffff7de7564 in _dl_catch_error (objname=objname@entry=0x7fffffffe150, errstring=errstring@entry=0x7fffffffe158,
    mallocedp=mallocedp@entry=0x7fffffffe14f, operate=operate@entry=0x7ffff7dec4d0 <dl_open_worker>, args=args@entry=0x7fffffffe160)
    at dl-error.c:187
#4  0x00007ffff7debda9 in _dl_open (file=0x4006e8 "/home/m/git/_uftrace/libmcount/libmcount-dynamic.so", mode=2,
    caller_dlopen=0x40062f <main+24>, nsid=-2, argc=<optimized out>, argv=<optimized out>, env=0x7fffffffe4a8) at dl-open.c:660
#5  0x00007ffff794c5ad in do_dlopen (ptr=ptr@entry=0x7fffffffe380) at dl-libc.c:87
#6  0x00007ffff7de7564 in _dl_catch_error (objname=0x7fffffffe370, errstring=0x7fffffffe378, mallocedp=0x7fffffffe36f,
    operate=0x7ffff794c570 <do_dlopen>, args=0x7fffffffe380) at dl-error.c:187
#7  0x00007ffff794c664 in dlerror_run (args=0x7fffffffe380, operate=0x7ffff794c570 <do_dlopen>) at dl-libc.c:46
#8  __GI___libc_dlopen_mode (name=<optimized out>, mode=<optimized out>) at dl-libc.c:163
#9  0x000000000040062f in main ()
```